lms/djangoapps/ccx/modulestore.py

```diff
@@ -0,0 +1,312 @@
+# -*- coding: utf-8 -*-
+"""A modulestore wrapper
+
+It will 'unwrap' ccx keys on the way in and re-wrap them on the way out
+
+In practical terms this means that when an object is retrieved from modulestore
+using a CCXLocator or CCXBlockUsageLocator as the key, the equivalent
+CourseLocator or BlockUsageLocator will actually be used. And all objects
+returned from the modulestore will have their keys updated to be the CCX
+version that was passed in.
+"""
+from contextlib import contextmanager
+from ccx_keys.locator import CCXLocator, CCXBlockUsageLocator
+from opaque_keys.edx.locator import CourseLocator, BlockUsageLocator
+
+
+def ccx_locator_to_course_locator(ccx_locator):
+    return CourseLocator(
+        org=ccx_locator.org,
+        course=ccx_locator.course,
+        run=ccx_locator.run,
+        branch=ccx_locator.branch,
+        version_guid=ccx_locator.version_guid,
+    )
+
+
+def strip_ccx(val):
+    retval = val
+    ccx_id = None
+    if hasattr(retval, 'ccx'):
+        if isinstance(retval, CCXLocator):
+            ccx_id = retval.ccx
+            retval = ccx_locator_to_course_locator(retval)
+        elif isinstance(object, CCXBlockUsageLocator):
+            ccx_locator = retval.course_key
+            ccx_id = ccx_locator.ccx
+            course_locator = ccx_locator_to_course_locator(ccx_locator)
+            retval = BlockUsageLocator(
+                course_locator, retval.block_type, retval.block_id
+            )
+    if hasattr(retval, 'location'):
+        retval.location, ccx_id = strip_ccx(retval.location)
+    return retval, ccx_id
+
+
+def restore_ccx(val, ccx_id):
+    if isinstance(val, CourseLocator):
+        return CCXLocator.from_course_locator(val, ccx_id)
+    elif isinstance(val, BlockUsageLocator):
+        ccx_key = restore_ccx(val.course_key, ccx_id)
+        val = CCXBlockUsageLocator(ccx_key, val.block_type, val.block_id)
+    if hasattr(val, 'location'):
+        val.location = restore_ccx(val.location, ccx_id)
+    if hasattr(val, 'children'):
+        val.children = restore_ccx_collection(val.children, ccx_id)
+    return val
+
+
+def restore_ccx_collection(field_value, ccx_id):
+    if isinstance(field_value, list):
+        field_value = [restore_ccx(fv, ccx_id) for fv in field_value]
+    elif isinstance(field_value, dict):
+        for key, val in field_value.iteritems():
+            field_value[key] = restore_ccx(val, ccx_id)
+    else:
+        field_value = restore_ccx(field_value, ccx_id)
+    return field_value
+
+
+class CCXModulestoreWrapper(object):
+
+    def __init__(self, modulestore):
+        self._modulestore = modulestore
+
+    def __getattr__(self, name):
+        """pass missing attributes through to _modulestore
+        """
+        return getattr(self._modulestore, name)
```

Comment:
Do I?  Are there places in the code where attributes of the modulestore are set or deleted?  If so, then I would certainly need that.  So far that hasn't appeared to be the case.